src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -23,34 +24,54 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
                     return element!;
                 }
             }
-            else
+            else if (source is IList<TSource> list)
             {
-                if (source is IList<TSource> list)
-                {
-                    return list[index];
-                }
-
-                if (index >= 0)
-                {
-                    using (IEnumerator<TSource> e = source.GetEnumerator())
-                    {
-                        while (e.MoveNext())
-                        {
-                            if (index == 0)
-                            {
-                                return e.Current;
-                            }
-
-                            index--;
-                        }
-                    }
-                }
+                return list[index];
+            }
+            else if (TryGetElement(source, index, out TSource? element))
+            {
+                return element;
             }
 
             ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index);
             return default;
         }
 
+        /// <summary>Returns the element at a specified index in a sequence.</summary>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="index">The index of the element to retrieve, which is either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="index" /> is outside the bounds of the <paramref name="source" /> sequence.
+        /// </exception>
+        /// <returns>The element at the specified position in the <paramref name="source" /> sequence.</returns>
+        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (!index.IsFromEnd)
+            {
+                return source.ElementAt(index.Value);
+            }
+
+            if (source.TryGetNonEnumeratedCount(out int count))
```

Comment:
I think this belongs in the `TryGetElementFromEnd` method.

Reply:
if (source.TryGetNonEnumeratedCount(out int count))
 - In ELementAt(Index), it calls source.ElementAt(int), which always calls list indexer.
 - In ElementAtOrDefault(Index), it calls source.ElementAtOrDefault(int), which does not call list indexer if index is out of bounds.
 
They work differently. So I didn't merge the `if(source.TryGetNonEnumeratedCount(out int count))` statement into `TryGetElementFromEnd`.

Reply:
Makes sense, thanks for clarifying.